In [3]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
%matplotlib inline

# For this homework is used cars dataset

In [4]:
df = pd.read_csv('./data/2022_cars.csv')

In [6]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [7]:
df.dtypes

Make                  object
Model                 object
Year                   int64
Engine Fuel Type      object
Engine HP            float64
Engine Cylinders     float64
Transmission Type     object
Driven_Wheels         object
Number of Doors      float64
Market Category       object
Vehicle Size          object
Vehicle Style         object
highway MPG            int64
city mpg               int64
Popularity             int64
MSRP                   int64
dtype: object

In [10]:
df.describe()

Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549.000000  2.100000e+04  
50%       26.000000     18.000000   1385.000000  2.999500e+04  
75%       30.000000     22.000000   2009.000000  4.223125e+04  
max      354.000000    137.000000   5657.000000  2.065902e+06

In [11]:
df.describe().round(2)

Year  Engine HP  Engine Cylinders  Number of Doors  highway MPG  \
count  11914.00   11845.00          11884.00         11908.00     11914.00   
mean    2010.38     249.39              5.63             3.44        26.64   
std        7.58     109.19              1.78             0.88         8.86   
min     1990.00      55.00              0.00             2.00        12.00   
25%     2007.00     170.00              4.00             2.00        22.00   
50%     2015.00     227.00              6.00             4.00        26.00   
75%     2016.00     300.00              6.00             4.00        30.00   
max     2017.00    1001.00             16.00             4.00       354.00   

       city mpg  Popularity        MSRP  
count  11914.00    11914.00    11914.00  
mean      19.73     1554.91    40594.74  
std        8.99     1441.86    60109.10  
min        7.00        2.00     2000.00  
25%       16.00      549.00    21000.00  
50%       18.00     1385.00    29995.00  
75%       22.00     2009.00    42231.25  
max      137.00     5657.00  2065902.00

# Question 1

What's the version of NumPy that you installed?

In [2]:
np.__version__

'2.3.1'

# Question 2

How many records are in the dataset?

Here you need to specify the number of rows.

- 16
- 6572
- 11914
- 18990

In [9]:
len(df)

11914

# Question 3

Who are the most frequent car manufacturers (top-3) according to the dataset?

- Chevrolet, Volkswagen, Toyota
- Chevrolet, Ford, Toyota
- Ford, Volkswagen, Toyota
- Chevrolet, Ford, Volkswagen

Note: You should rely on "Make" column in this question.

In [17]:
df.groupby('Make')

# Question 4

What's the number of unique Audi car models in the dataset?

- 3
- 16
- 26
- 34

# Question 5

How many columns in the dataset have missing values?

- 5
- 6
- 7
- 8

# Question 6

Find the median value of "Engine Cylinders" column in the dataset.
Next, calculate the most frequent value of the same "Engine Cylinders".
Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
Now, calculate the median value of "Engine Cylinders" once again.
Has it changed?

Hint: refer to existing mode and median functions to complete the task.

- Yes
- No

# Question 7
Select all the "Lotus" cars from the dataset.
Select only columns "Engine HP", "Engine Cylinders".
Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
Get the underlying NumPy array. Let's call it X.
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
Compute the inverse of XTX.
Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the value of the first element of w?
Note: You just implemented linear regression. We'll talk about it in the next lesson.

- -0.0723
- 4.5949
- 31.6537
- 63.5643